ember-popovers

a15n opened the pull request:
## ember-popovers
![popovers](https://cloud.githubusercontent.com/assets/8782940/17453888/276f6a5e-5b3a-11e6-8683-3c45fc017e7d.jpg)

This PR addresses the need for a popover expressed here (https://github.com/sir-dunxalot/ember-tooltips/issues/77) and implied here (https://github.com/sir-dunxalot/ember-tooltips/issues/51)

* creates `popover-on-element` and `popover-on-component` components which extend their tooltip counterparts
  * these all have the same [API options as ember tooltips](https://github.com/sir-dunxalot/ember-tooltips#options)
  * these behave the exact same way EXCEPT
    * event handling is custom and mirrors that found in Twitter popovers.
    * a new `hideDelay` param which defines the milliseconds a user may hover outside of the popover before it hides. This is necessary because popovers can be spaced N pixels away from their target element
    * popovers have no visual CSS properties. These must be defined by the user
* adds popover specific tests in the `tests/integration/components/popover/` directory
* exposes a `hide` action for use within the popover's block (ie clicking an interior close button)

### Exhaustive list of changes made to tooltip-on-element
* `git mv tooltip-on-element.js tooltip-and-popover.js` (https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/c5da8b755e141b49cc3b0cb64b71ece927da5527)
* moved the tooltip-specific event handling into tooltip-on-element's didInsertElement (https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/2b99f85f041003a3bfb5fa227afd41af196e1bf3)
* use `classPrefix` instead of one-off strings. Change classPrefix to `tooltip-and-popover` (https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/9dc2fc76909c5560b9be16b24d97c4a2ce148cba)
* used `classPrefix: 'ember-tooltip-or-popover',` (https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/1e9cef210184256810ee37795292f45af1ffa956 and https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/5045c4c5e8818690c70b163b6fdf81ae86072057)
* deprecated `tooltipIsVisible` in favor of `isShown` (https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/110c0fa2ac1198cb3f846e82264d32d9e27a0ac3)
* deprecated `onTooltipShow` in favor of `onShow`, etc (https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/599399719575a69686473dc0ca38ad56d76375f6)
* replaced `tooltipCounterId` with `tooltipOrPopoverCounterId` (https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/26fcdd49fdf87f4d9df5c83b9f5f27bbeb5e394f)

### Popover event="click"
* clicking the target element will toggle the popover
* clicking inside the popover will not toggle it
* clicking outside of the target and the popover will hide it
![popover-click](https://cloud.githubusercontent.com/assets/7050871/18113143/4e29bd32-6ee2-11e6-88e1-3418fa4150f8.gif)

### Popover event="hover" (not suitable for mobile)
* hovering on the target element will show the popover
* hovering away from the target element will hide the popover after `hideDelay` milliseconds (default 250)
* hovering from the target element to the popover works if it happens during the `hideDelay`
![popover-hover](https://cloud.githubusercontent.com/assets/7050871/18113238/e010ee64-6ee2-11e6-9ff1-a0c674a6d702.gif)

### Popover focus (available when event="focus", "click", or "hover"
* focusing into the target will show the popover
* focusing into the popover will keep the popover visible
* focusing out of the popover and target will hide the popover
![popover-focus](https://cloud.githubusercontent.com/assets/7050871/18216215/50fc3c5e-710a-11e6-909f-3b65e094f0c9.gif)


### Popover hide API
* invoke the popover like so and gain access to the `hide` action
```
{{#popover-on-element as |popover|}}
	Click <a href {{action popover.hide}}>here</a> to hide popover
{{/popover-on-element}}
```
![popover-hide](https://cloud.githubusercontent.com/assets/7050871/18113315/8d8c7086-6ee3-11e6-858c-29a44c3d2dd3.gif)

### TODO (Blocking)
- [x] simplify the didInsertElement _showOn/_hideOn functionality if possible
- [x] rename tooltip-and-popover.tooltipIsVisible to `isVisible` 
- [x] document all any differences in tooltip-and-popover made after the git mv
- [x] document how `{{action popover.hide}}` works and why the package.json changes were needed
- [x] fix display issue when effect="fade", spacing="0", event="hover"
- [x] fix display issue when effect="fade", spacing="0", event="click"
- [x] add gifs showing the interaction
- [x] get feedback on appropriate show/hide interactions to include/exclude, and write comprehensive tests for them
- [x] add `event='focus'` handling
- [x] get tests passing in separate PR first
- [x] confirm the correct class name paradigm (currently tooltip-and-popover, ember-tooltip, and ember-popover)
- [x] replace `tooltipIsVisible` with something more generic (not isVisible, maybe isShown or visible)

a15n commented:
even with a simple change, tests are failing...

https://travis-ci.org/sir-dunxalot/ember-tooltips/jobs/156037676

MiguelMadero commented on addon/templates/components/popover-on-element.hbs:
```diff
@@ -0,0 +1,3 @@
+{{yield (hash
+	hide=(action "hide")
```
minor: indent

a15n replied:
this is a github display issue. It is correctly indented in my IDE

MiguelMadero commented on addon/components/popover-on-element.js:
```diff
@@ -1,13 +1,14 @@
 
   hideDelay: '250',
 
+  template,
```
you need to use layout instead of template, template is the block, layout is what yields

a15n commented on package.json:
```diff
@@ -54,7 +54,9 @@
-    "ember-cli-sass": "5.3.1"
+    "ember-cli-htmlbars": "^1.1.0",
+    "ember-cli-sass": "5.3.1",
+    "ember-hash-helper-polyfill": "^0.1.1"
```
we need `ember-hash-helper-polyfill` and `ember-cli-htmlbars` because of the (hash) template that passes the `hide` closure action to the yield

a15n replied:
ember-hash-helper-polyfill is only needed till Ember 2.3

sir-dunxalot commented:
Hey @a15n - any idea why the tests are failing on the initial commit? Could it be as simple as merging with `master`?

a15n commented:
@sir-dunxalot I made sure to use the most up to date version before I started. I'm checking my `git log` and I have the 2.0.1 release and all the subsequent PRs. When I run the tests locally I get the following errors. These are the same errors I noticed when I tested the 'initial commit'.

![screen shot 2016-08-30 at 9 33 12 am](https://cloud.githubusercontent.com/assets/7050871/18097931/e1cd8ea4-6e94-11e6-9115-5fd3b499879d.png)
----------
![screen shot 2016-08-30 at 9 32 59 am](https://cloud.githubusercontent.com/assets/7050871/18097930/e1c9b66c-6e94-11e6-8d03-b413d14df3b9.png)
----------
![screen shot 2016-08-30 at 9 33 24 am](https://cloud.githubusercontent.com/assets/7050871/18097929/e1b51b58-6e94-11e6-8653-9816117c3b66.png)

a15n commented:
Also...

![screen shot 2016-08-30 at 9 36 22 am](https://cloud.githubusercontent.com/assets/7050871/18098008/40a6c1ac-6e95-11e6-80ee-2017bbe20935.png)

sir-dunxalot commented:
Hmm, ok. I think either you or I will have to work on this branch and get it passing before merging in. With a PR this big I'd rather not merge onto master and then get things to pass (incase others are pulling down master into their projects directly). I could pull down from your branch if preferred.

Alternatively, we could reissue this PR to a `popover` branch on this repo and I could try and get the tests to pass on that before merging it into `master`.

Let me know how you would like to proceed. I know you've done tons of work on this so I don't want to have you to put in more work especially if it's unrelated to this feature.

a15n commented:
> "With a PR this big I'd rather not merge onto master and then get things to pass (incase others are pulling down master into their projects directly)"

Agreed. If you were able to pull down this branch and investigate that would be great. There are popover tests failing right now but I am in the process of fixing those. 

Thanks

a15n commented:
@sir-dunxalot regarding class name structure I have `.ember-tooltip` (no changes), `.ember-popover`, and `.tooltip-and-popover` for the CSS attributes that are shared by both. What do you think would be best?

MiguelMadero commented on addon/components/popover-on-element.js:
```diff
@@ -40,25 +40,28 @@ export default TooltipAndPopoverComponent.extend({
     } else if (_showOn === 'click') {
 
-      $target.on(_showOn, (event) => {
+      $(document).on('click', (event) => {
```
Lets make sure you unsubscribe from all events, but specially this one.

a15n commented on addon/components/popover-on-element.js:
```diff
@@ -64,6 +65,42 @@ export default TooltipAndPopoverComponent.extend({
+      return;
+    }
+
+    if (_showOn === 'mouseenter') {
```
I think I'll be able to simplify the `_showOn == 'mouseenter'` case now that I only use _isMouseInside for that approach.

a15n commented:
@sir-dunxalot the only remaining, blocking items are 

- [ ] get tests passing in separate PR first **(Duncan are you still able to help with this?)**
- [ ] confirm the correct class name paradigm (currently tooltip-and-popover, ember-tooltip, and ember-popover) **(how do you feel about this)**
- [ ] replace tooltipIsVisible with something more generic **(I tried to use `isVisible` but ran into problems (I think it may be reserved). If that doesn't work how do you feel about `isShown`, `isTooltipOrPopoverVisible`, or `visible`?)**

There is one TODO I have listed as non-blocking which is adding "focus" events to the popover. I attempted that but found it very difficult! I'll document my findings on it.

ZSujay commented on addon/components/popover-on-element.js:
```diff
@@ -0,0 +1,95 @@
+        // if a click is NOT on $target and NOT an ancestor of $target.
+        // If so then it must be a click elsewhere and should close the popover
+        // see... https://css-tricks.com/dangers-stopping-event-propagation/
+        const $clickedElement = event.target;
```
@a15n In general don't use the `$` prefix for naming variables/constants unless they're jQuery-fied.

ZSujay commented on addon/components/popover-on-element.js:
```diff
@@ -0,0 +1,95 @@
+import Ember from 'ember';
+import TooltipAndPopoverComponent from 'ember-tooltips/components/tooltip-and-popover';
```
Changes to this file LGTM except for the minor comment

a15n commented:
@sir-dunxalot great news. I was able to figure out a way to cleanly support focus events. Here is the commit for it (https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/8653877b56ab3c5e8da6e1b6c7e23f607c15c33f)

Now that this is done the final things to do are (1) confirm className paradigm, replace tooltipIsVisible, and (3) get tests passing. 

I'll make a fresh PR and see if I can get the tests passing. If I can figure a way out I'll use it here.

a15n commented:
@sir-dunxalot, @MiguelMadero boom, tests are passing. I don't know what introduced this issue but the problem was that the failing tests all had two `this.render()` in them. I split them into individual tests and it fixed it.

https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/f6a4632b3d61be408ef305fa1623f1b19822c44b

sir-dunxalot commented:
Oh man, great work! You beat me to it - I owe you a beer if you're ever in Boston!

The only class name I'm unsure about is `.tooltip-and-popover` because the others are prefixed. At minimum, I think this should be `ember-tooltip-or-popover` or something like that. Thoughts?

As for `tooltipIsVisible`, you are correct in saying `isVisible` is reserved by Ember. I like `isShown` if we want to make it more generic.

a15n commented:
That would be great! And if you ever come out to SF we should meet up for a drink.

I think that `.ember-tooltip-or-popover` is a good class name. And `isShown` was my top choice too.

I have also updated some other tooltip-specific names to their generic counterpart. What do you think of
* [tooltipCounterId -> tooltipOrPopoverCounterId](https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/26fcdd49fdf87f4d9df5c83b9f5f27bbeb5e394f)
* [onTooltipShow -> onShow, etc](https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/4a7303d1bc87b662b1612aa39d9aaa4267e54672)

I believe that is everything for this PR! It's a pretty big change that will probably require an update of the readme, demo page, and a v3 version. Let me know if I can help with any of that.

sir-dunxalot commented:
Will do - I'll let you know!

I agree with all the above changes.

We will definitely need updating of all the resources you mentioned too. I'm happy to do it but if you're volunteering of course that would be great too. I will plan sometime this weekend to document and release this if we don't get round to it sooner.

It sounds like the code-portion of this is all set now?

a15n commented:
Yes the code portion is all set now. Things have gotten rather busy for me at work so I don't think I'll be able to help with updating the resources. 

We are good to wait till this weekend. Thanks!

sir-dunxalot commented:
Ok, great. I'll wait to merge this in until I have the documentation in place. I don't mind a 10-minute overlap of merging without the docs updated so I'll do it on the fly. I'll also do a QA run through - will keep you updated.

jamesdixon commented:
Well done, gents! Looking forward to trying this out.

andrew-zenefits commented:
@sir-dunxalot you'll notice in this latest PR that I have improved the jQuery checks to determine if an element is inside the popover or elsewhere. I tested out the new popovers in our code base and learned that sometimes all of the tethers are contained within a div at the bottom, rather than within the target. For this reason the following code didn't work and this fix was necessary..

```
$target.find(clickedElement).length;
// didn't work because sometimes the popover was outside of the target
```

https://github.com/sir-dunxalot/ember-tooltips/pull/92/commits/3248b18a71d18e8b3c9621d89cea9c3006b8be9f

jliuzen commented on addon/components/popover-on-element.js:
```diff
@@ -0,0 +1,120 @@
+  hideDelay: '250',
+
+  layout,
+  classNames: ['ember-popover'],
```
nit/personal preference: put `classNames` at top

jliuzen commented on addon/components/popover-on-element.js:
```diff
@@ -0,0 +1,120 @@
+
+export default TooltipAndPopoverComponent.extend({
+
+  hideDelay: '250',
```
should this be an in int instead of string?

andrew-zenefits commented:
@sir-dunxalot, @MiguelMadero you'll notice in the last commit I've triggered every `event='focus'` test event with javascript instead of jQuery. While investigating the flaky `event='focus'` tests we discovered that each test would fail the second time it was called, *unless* the chrome screen was clicked just before the test was run. We suspected that the trigger('focus') events weren't being triggered correctly the second time. Miguel mentioned something about synthetic events being introduced in Ember 2.8. We think that Travis is being affected by the same thing.

jamesdixon commented:
How's this going guys? Anxious to give it a shot!

andrew-zenefits commented:
@sir-dunxalot is there anything I can do to help this land?

a15n commented:
@sir-dunxalot I'm glad I was able to work on this some more. I added backwards compatibility for `tooltipIsVisible` and the `onTooltipShow`, etc parameters. I have added tests to support these properties. I recommend we plan on removing support for these properties in v3.0.0.

I have also updated the readme to include popovers.

I believe this is ready to merge, BUT it still has flaky tests in `ember-beta`. I'm at a loss for how to address the flaky tests because I cannot get them to fail locally.

a15n commented:
@sir-dunxalot I've addressed some of the failing tests here. https://github.com/sir-dunxalot/ember-tooltips/pull/97

Let me know if I can help in any other way to land this.

sir-dunxalot commented:
Andrew, I'm finally getting around to this. This has been an immense effort - great work, truly.

#97 is now on master. It looks like that has caused merge conflicts. If we can fix those so that there's only the one test failure (the `el.classList is not an object` error) then I'm happy to merge this, fix that one test locally, and release).

a15n commented:
@sir-dunxalot I have merged with the newest master, resolved conflicts, and all tests are passing except the `Integration | Component | tooltip on component` tests on ember-beta and ember-canary. I think we're good to go?

sir-dunxalot commented:
Alright, I'm merging and pulling down now. I will try and get those last two test passing before releasing.

sir-dunxalot commented:
I've got four tests failing locally with the following error:

```
'[object EventConstructor]' is not a constructor (evaluating 'new window.Event('focus')')
```

I'm attempting to fix these before fixing the remaining two `tooltip on component` failing tests.
